File: test_pykokkos/mylib.py
"""
PyKokkos-based array library - mirrors the Kokkos mylib interface.

PyKokkos provides Kokkos-like parallel programming in Python with JIT compilation.

Install (requires conda or building from source):
    conda install -c conda-forge pykokkos
    # or
    pip install git+https://github.com/kokkos/pykokkos.git

Note: PyKokkos requires pykokkos-base which compiles Kokkos C++ code.
"""

import pykokkos as pk
from typing import List

# ============================================================================
# Initialization
# ============================================================================

_initialized = False


def initialize():
    """Initialize PyKokkos runtime."""
    global _initialized
    pk.set_default_space(pk.OpenMP)  # or pk.Cuda for GPU
    _initialized = True


def is_initialized() -> bool:
    """Check if PyKokkos is initialized."""
    return _initialized


# ============================================================================
# Array1D class - mirrors Kokkos Array1D interface
# ============================================================================

class Array1D:
    """1D array wrapper using PyKokkos View internally."""

    def __init__(self, n: int):
        self._size = n
        self._data = pk.View([n], dtype=pk.double)

    def size(self) -> int:
        return self._size

    def get(self, i: int) -> float:
        return float(self._data[i])

    def set(self, i: int, val: float):
        self._data[i] = val

    def from_list(self, values: List[float]):
        if len(values) != self._size:
            raise RuntimeError("Size mismatch")
        for i, v in enumerate(values):
            self._data[i] = v

    def to_list(self) -> List[float]:
        return [float(self._data[i]) for i in range(self._size)]

    @property
    def data(self):
        """Access underlying PyKokkos View (for kernels)."""
        return self._data

    def __len__(self) -> int:
        return self._size

    def __getitem__(self, i: int) -> float:
        return self.get(i)

    def __setitem__(self, i: int, val: float):
        self.set(i, val)


# ============================================================================
# Kernels
# ============================================================================

@pk.workunit
def _dot_kernel(i: int, acc: pk.Acc[pk.double], a: pk.View1D[pk.double], b: pk.View1D[pk.double]):
    """PyKokkos parallel_reduce kernel for dot product."""
    acc += a[i] * b[i]


def dot_product(a: Array1D, b: Array1D) -> float:
    """Compute dot product of two arrays (parallel via PyKokkos)."""
    if a.size() != b.size():
        raise RuntimeError("Size mismatch in dot_product")

    result = pk.parallel_reduce(a.size(), _dot_kernel, a=a.data, b=b.data)
    return float(result)


# Warmup function
def warmup(n: int = 1000):
    """Pre-compile kernels with dummy data."""
    a = Array1D(n)
    b = Array1D(n)
    a.from_list([1.0] * n)
    b.from_list([1.0] * n)
    _ = dot_product(a, b)

File: test_binder/src/mylib.cpp
#include "mylib.hpp"
#include <stdexcept>

// ============================================================================
// Kokkos initialization
// ============================================================================

void initialize_kokkos() {
    if (!Kokkos::is_initialized()) {
        Kokkos::initialize();
    }
}

bool is_kokkos_initialized() {
    return Kokkos::is_initialized();
}

// ============================================================================
// Array1D implementation
// ============================================================================

Array1D::Array1D(size_t n) : data_("array", n), size_(n) {
    host_data_ = Kokkos::create_mirror_view(data_);
}

size_t Array1D::size() const {
    return size_;
}

void Array1D::set(size_t i, double val) {
    if (i >= size_) throw std::out_of_range("Index out of bounds");
    host_data_(i) = val;
}

double Array1D::get(size_t i) const {
    if (i >= size_) throw std::out_of_range("Index out of bounds");
    return host_data_(i);
}

void Array1D::sync_to_device() {
    Kokkos::deep_copy(data_, host_data_);
}

void Array1D::sync_to_host() {
    Kokkos::deep_copy(host_data_, data_);
}

void Array1D::from_list(const std::vector<double>& values) {
    if (values.size() != size_) {
        throw std::runtime_error("Size mismatch");
    }
    for (size_t i = 0; i < size_; ++i) {
        host_data_(i) = values[i];
    }
    sync_to_device();
}

std::vector<double> Array1D::to_list() const {
    std::vector<double> result(size_);
    for (size_t i = 0; i < size_; ++i) {
        result[i] = host_data_(i);
    }
    return result;
}

Array1D::ViewType& Array1D::view() {
    return data_;
}

const Array1D::ViewType& Array1D::view() const {
    return data_;
}

// ============================================================================
// Kernel implementation
// ============================================================================

double dot_product(const Array1D& a, const Array1D& b) {
    if (a.size() != b.size()) {
        throw std::runtime_error("Size mismatch in dot_product");
    }

    auto va = a.view();
    auto vb = b.view();
    size_t n = a.size();
    double result = 0.0;

    Kokkos::parallel_reduce("dot_product", n,
        KOKKOS_LAMBDA(const int i, double& lsum) {
            lsum += va(i) * vb(i);
        },
        result
    );

    return result;
}
